### PR TITLE
added domains

### PR DESCRIPTION
**Pull request by Felalolf**
These are the super secret domains. Using a domain should output a warning, which is not implemented at this moment in time.

### CODE REVIEW

File: src/main/scala/viper/gobra/ast/frontend/Ast.scala [hunk 1]
```diff
@@ -1038,6 +1037,19 @@ case class PMultisetType(elem : PType) extends PGhostLiteralType
 /** The type of option types. */
 case class POptionType(elem : PType) extends PGhostLiteralType
 
+
+case class PDomainType(funcs: Vector[PDomainFunction], axioms: Vector[PDomainAxiom]) extends PGhostLiteralType with PUnorderedScope
```
Comment (jcp19) [inline src/main/scala/viper/gobra/ast/frontend/Ast.scala:1041]: Is there any reason why you disallowed parametric domains?

File: src/main/scala/viper/gobra/frontend/Desugar.scala [hunk 2]
```diff
@@ -1952,13 +1968,66 @@ object Desugar {
...
+        additionalMembers
+      }
+    }
+
+
+    def registerDomain(t: Type.DomainT, dT: in.DomainT): Unit = {
+      if (!registeredDomains.contains(dT.name) && info == t.context.getTypeInfo) {
+        registeredDomains += dT.name
+
+        AdditionalMembers.addFinalizingComputation{ () =>
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Desugar.scala:2004]: why are these computations delayed to the finalizing step?

File: src/main/scala/viper/gobra/frontend/Desugar.scala [hunk 3]
```diff
@@ -1952,13 +1968,66 @@ object Desugar {
...
+          finalized = true
+        }
+
+        additionalMembers
+      }
+    }
+
+
+    def registerDomain(t: Type.DomainT, dT: in.DomainT): Unit = {
+      if (!registeredDomains.contains(dT.name) && info == t.context.getTypeInfo) {
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Desugar.scala:2001]: Is `info == t.context.getTypeInfo` a sanity check? Shouldn't an exception be thrown if `info` does not match?

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 4]
```diff
@@ -965,6 +965,17 @@ object Parser {
     lazy val optionType : Parser[POptionType] =
       "option" ~> ("[" ~> typ <~ "]") ^^ POptionType
 
+    lazy val domainType: Parser[PDomainType] =
+      "domain" ~> "{" ~> (domainClause <~ eos).* <~ "}" ^^ { clauses =>
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Parser.scala:969]:
```suggestion
      "domain" ~> "{" ~> repsep(domainClause, eos) <~ eos.? <~ "}" ^^ { clauses =>
```

File: src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala [hunk 5]
```diff
@@ -0,0 +1,101 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala:5]:
```suggestion
// Copyright (c) 2011-2021 ETH Zurich.
```

File: src/main/scala/viper/gobra/ast/frontend/Ast.scala (hunk 1 above)
Comment (jcp19) [inline src/main/scala/viper/gobra/ast/frontend/Ast.scala:1041]: Just to be sure, you decided not to support parametric domains to be consistent with Go's lack of support for generics right?

**Review by jcp19 — APPROVED** [review]
It looks fine to me

File: src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      val (dPos, dInfo, dErrT) = domain.vprMeta
+
+      val funcs = domain.funcs map { f =>
+        val (fPos, fInfo, fErrT) = f.vprMeta
+        val formalArgs = f.args map ctx.typeEncoding.variable(ctx)
+        val resType = ctx.typeEncoding.typ(ctx)(f.results.typ)
+        vpr.DomainFunc(f.name.name, formalArgs, resType)(fPos, fInfo, domainName, fErrT)
+      }
+
+      val defaultFunc = vpr.DomainFunc(
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala:48]: It seems a bit weird to me that there is always a default function generated under the hood for every domain. Do we need this? Does this lead to weird side effects that one cannot state disequalities between default function and a user-defined constructor?

**Review by ArquintL — APPROVED** [review]
I only have some minor comments other than that it looks good to me

File: src/main/scala/viper/gobra/ast/frontend/Ast.scala (hunk 1 above)
Comment (ArquintL) [reply src/main/scala/viper/gobra/ast/frontend/Ast.scala:1041]: We do not need generics right now and eventually there will be true ADT support

File: src/main/scala/viper/gobra/frontend/Desugar.scala (hunk 2 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Desugar.scala:2004]: To avoid cylces.

File: src/main/scala/viper/gobra/frontend/Desugar.scala (hunk 3 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Desugar.scala:2001]: No, that is not a sanity check. This function can be called from the context of another file.

File: src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala (hunk 6 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala:48]: In the current infrastructure, everything must have a default value. I do not know what you mean with "disequalities between default function and a user-defined constructor".

**Comment by jcp19** [comment]
Is there any reason keeping us from merging this PR?

**Pull request merged**